### PR TITLE
feat(dbt): use `TableMetadataEntries` to emit column lineage

### PR DESCRIPTION
**Pull request by rexledesma**
## Summary & Motivation
Apply https://github.com/dagster-io/dagster/pull/20477 on https://github.com/dagster-io/dagster/pull/20407 to see how the API feels.

## How I Tested These Changes
pytest

### CODE REVIEW

**Comment by rexledesma** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20569?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#20569** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20569?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#20477** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20477?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#20548** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20548?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#20407** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20407?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#20531** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20531?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rexledesma and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 1]
```diff
@@ -398,26 +395,16 @@ def _build_column_lineage_metadata(
...
 
-                    parent_columns.add((parent_asset_key, column.name))
+                    column_deps.append(
+                        TableColumnDep(asset_key=parent_asset_key, column_name=column.name)
+                    )
 
-            lineage_metadata[column_name] = list(parent_columns)
+            deps_by_column[column_name] = column_deps
 
         # 4. Render the lineage as a JSON blob.
```
Comment (sryza) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:404]: I think this comment is now inaccurate

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 2]
```diff
@@ -378,8 +376,7 @@ def _build_column_lineage_metadata(
         column_names = cast(exp.Query, optimized_node_ast).named_selects
 
         # 3. For each column, retrieve its dependencies on upstream columns from direct parents.
-        # TODO: this should be refactored as a Python object that renders the JSON metadata.
-        lineage_metadata: Dict[str, List[Tuple[AssetKey, str]]] = {}
+        deps_by_column: Mapping[str, Sequence[TableColumnDep]] = {}
```
Comment (sryza) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:379]: I think this needs to be a `Dict` because it's mutated below

**Review by sryza — APPROVED** [review]
LGTM after addressing my small comments

**Comment by rexledesma** [comment]
### Merge activity

* **Mar 20, 4:11 PM PDT**: @rexledesma started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/20569).

**Pull request merged**